src/content/posts/4-way-handshake.md/index.md

```diff
@@ -0,0 +1,139 @@
+
+**VERY IMPORTANT**
+
+*For WPA-Entreprise and WPA2-Entreprise, they use here the PMK created by RADIUS authentication*
+
+## Second Request
+Let’s make a short recap.  
+At the beginning, we only know the **PSK=PMK**, which we derive from the **network password** and **SSID**.  
+*PMK is derived from RADIUS authentication in case of WPA-Entreprise and WPA2-Entreprise*  
+
+We received the AP's nonce from the first request, and the station generated the SNonce to create the **PTK** from the **ANonce**, **SNonce**, and **PSK**.
+
+Now we need to tell the AP that we know the password without explicitly sending the password.
+
+The station sends a packet containing the SNonce, allowing the AP to generate the same PTK.
+At the end of the packet, a **signature called the Message Integrity Code (MIC)** is added.
+  
+The MIC is just a SHA-1 hash of the packet content with the PTK as salt.  
+When the AP receives the second request, it will generate its PTK (it has all the necessary information: SNonce, ANonce, PSK).
+However, at this moment, the AP doesn't know if the station has the correct password.
+
+The AP generates the MIC from its newly generated PTK and compares it with the provided MIC.  
+**If the MICs are identical => PTKs are identical => PSKs are identical => passwords are identical**
+
+## Third Request
+To tell the station that it has the correct password, the AP won't do so explicitly.  
+**The AP will use the same strategy: a MIC.**
+
+To optimize the request, the AP will send useful information: the Group Temporal Key (GTK), directly generated by the AP.
+
+To check if the station has the correct password, the AP generates the MIC corresponding to the GTK (packet content) using the PTK.
+
+If both match, the station and AP are sure that the other has the correct password.
+
+## Last Request
+The station sends an acknowledgment packet to tell the AP that everything is OK.
+
+**The encrypted communication can now begin.**
+
+## Encryption
+WPA2 uses **AES** while WPA uses **TKIP** to encrypt packet content.
+Even if both are encryption's methods, AES is more secure.
+This makes it impossible for an hacker to read a captured packet.
+
+The PTK contains several keys.  
+Several keys are derived from the PTK:
+1. **KCK** = PTK bits 0-127, the key confirmation key -> used for MIC (verifying message integrity).
+2. **KEK** = PTK bits 128-255, the key encryption key -> used for encrypting the key data in the authentication packets.
+3. **TEK** = PTK bits 256-383, the temporal encryption key for unicast traffic -> used for AES encryption.
+
+We simplify things by saying that we use the PTK for something, but in reality, we often use just a part of it.
+
+## Recap of Key Derivation
+![key_derivation](key_derivation.png)
+
+In WPA2-PSK and WPA-PSK, the PMK is the same as the PSK.
+
+## Conclusion
+Remember that even if we use WPA/WPA2, if a hacker catch all packets, he can brutforce PTK and get WiFi.

```

Comment:
he can bruteforce the PTK and access the WiFi. Okay, but how hard it actually is to bruteforce ? Here, it would be interesting to 1) give recommendations as of the length/ complexity required for the password to make it very unlikely to be broken (for instance, probability that a hacker with a good config can break it within 24 hours < 10^-3), 2) show the difference between WPA and WPA2 for this matter , graphs would be perzct

Reply:
good recommandation but this subject is so interessant that we should make an complete article about this and add link at the end of this post